pyo3-ffi/src/object.rs

```diff
@@ -72,99 +104,85 @@ pub unsafe fn Py_TYPE(ob: *mut PyObject) -> *mut PyTypeObject {
     (*ob).ob_type
 }
 
+// PyLong_Type defined in longobject.rs
+// PyBool_Type defined in boolobject.rs
+
 #[inline]
 pub unsafe fn Py_SIZE(ob: *mut PyObject) -> Py_ssize_t {
-    (*(ob as *mut PyVarObject)).ob_size
+    debug_assert_ne!((*ob).ob_type, std::ptr::addr_of_mut!(crate::PyLong_Type));
+    debug_assert_ne!((*ob).ob_type, std::ptr::addr_of_mut!(crate::PyBool_Type));
+    (*ob.cast::<PyVarObject>()).ob_size
 }
 
 #[inline]
 pub unsafe fn Py_IS_TYPE(ob: *mut PyObject, tp: *mut PyTypeObject) -> c_int {
     (Py_TYPE(ob) == tp) as c_int
 }
 
+#[inline(always)]
+#[cfg(all(Py_3_12, not(target_pointer_width = "32")))]
+pub unsafe fn _Py_IsImmortal(op: *mut PyObject) -> c_int {
+    (((*op).ob_refcnt.ob_refcnt as crate::PY_INT32_T) < 0) as c_int
+}
+
+#[inline(always)]
+#[cfg(all(Py_3_12, target_pointer_width = "32"))]
+pub unsafe fn _Py_IsImmortal(op: *mut PyObject) -> c_int {
+    ((*op).ob_refcnt.ob_refcnt == _Py_IMMORTAL_REFCNT) as c_int
```

Comment:
Are the top two bits never modified for immortal objects?

Reply:
This implementation matches the C header: https://github.com/python/cpython/blob/e3b5ed7b1cb8ad12aa285cf78303c6dda0953181/Include/object.h#L245

I think in practice the top two bits are reserved for abi3 builds built against older Python implementations, for which `Py_INCREF` would still directly modify the refcount field and so immortal objects need to allow some space for the count to spill over without wrapping.

Reply:
Ok, the salient point here is probably

> The object would lose its immortality, but the execution would still be correct.

from the C header. So indeed, in 32-bit builds, these object would really not be considered "immortal" any more after such an extension touched their reference.

Weird, but matching upstream behaviour. (I really wonder whether the benefit of having immortal objects is worth this technical complexity.)